near-wallet/1775 - Ledger transport fallback support

MaximusHaximus opened the pull request:
Re: https://github.com/near/near-wallet/issues/1775

No directly breaking changes were made to the existing `createClient()` method, which _takes a transport as an argument_

Instead, we've added a new method to the exports from the `index.js` module: **`getSupportedTransport()`**
- This method uses methods provided by LedgerHQ's WebHID, WebUSB, and finally U2F transports to identify which transport is supported by the current browser to create a connection to the Ledger device.
- If an attempt to create what we _think_ is a supported transport fails (e.g. the transport advertises that it is supported on the current browser, but initializing that transport throws an error), then we fall back to trying a U2F transport, which is what we used prior to this PR.

Consumers that want to use our fallback rather than providing an explicit transport need only use e.g.:
```javascript
  import { createClient, getSupportedTransport } from "near-ledger-js";
  
  const transport = await getSupportedTransport();
  transport.setScrambleKey("NEAR");
  const client = await createClient(transport);
```

NOTE:
1. Both WebHID & WebUSB require they are initialized _as a result of a user action_.  This may mean that using these transports could be a breaking change if the consumer of this code was previously creating a ledger client passively when their page was being rendered (this is how the wallet did things previously).  The only way to ensure that all transports will work is for the client creation to happen in response to an e.g. click event from the user.
- So I added a 'Connect to Ledger' button to the demo page.
2. If you launch the NEAR app after connecting to the ledger, or if you quit the NEAR app after connecting to the ledger, then the connection will be closed.
- So I added tracking of connected status to the demo HTML UI by adding an `on('disconnect'...)` handler to the ledger transport and updating the UI whenever a connect or disconnect action happens (see demo.js)

mikedotexe commented:
Happy to have my team code review as soon as this is no longer a work in progress.

ilblackdragon commented:
This is really pressing issue for large token holders. What are missing to finalize this PR?

MaximusHaximus commented:
@ilblackdragon These changes are ready for review; I've tested them on OS X on both Chrome, Firefox, and Brave.

Any apps that use this will need to be updated after we've got this merged and published to NPM.  Per the details on the description for this PR, the new transports must be initialized as a result of a user action, whereas `U2F` could be initialized in the background (e.g. on page render).

This change means that not only will any UIs that use this package need to be updated to use the new helper function from this PR to create their transport, but they may need to be refactored to move initialization of the Ledger into an e.g. onClick event if they were previously doing it in the background (this includes our own wallet frontend, which currently does this initialization in the background)

MaximusHaximus commented:
@mikedotexe Review/Hammer away! Some extra testing from multiple folks with real transactions just to make sure the new transport is stable would be a good idea I think

volovyk-s reviewed (commented):
Please, check this comment: https://github.com/near/near-wallet/issues/1775#issuecomment-859571514

MaximusHaximus commented on index.js:
```diff
@@ -1,3 +1,12 @@
+const { default: LedgerTransportWebHid } = require('@ledgerhq/hw-transport-webhid');
+const platform = require('platform');
+
+const ENABLE_DEBUG_LOGGING = true;
```
:+1: updated to be defaulted to disabled in `index.js`. Added an export to control it externally, `setDebugLogging()`, and modified `demo.js` to use that method to enable debug logging in the demo itself :)

MaximusHaximus commented:
@volovyk-s I modified this code to be more robust in the case of errors being encountered while actually trying to call `create()` on a transport, even though `isSupported()` for that transport returned `true`.  I did this because in my testing, `isSupported()` seems to often return true in cases where it is actually broken if you try to use that transport.  The new fallback will try to call `create()` on the preferred transport (WebHID), and if an error is encountered, will try the other transports that are considered supported due to the result of calling`isSupported()` on them in sequence.

If all transports throw errors during attempts to call `.create()`, then the `getSupportedTransport()` method will throw.

I also moved the `getSupportedTransport()` logic and dependencies into their own file so that the transports from `@ledgerhq` could be tree shaken out for people who use this package but don't want to bring in all 3 transports as dependencies because they are writing something that targets one-and-only one transport that they provide to the constructor of `createClient()`

I also enhanced state management for creating new transports after fatal errors have been encountered with an existing transport and encapsulated the state into a class in `demo.js`.

I also moved some stuff off of `window` and into `window.nearLedger` to avoid shadowing function names and variables that were incidentally the same.

volovyk-s commented on supportedTransports.js:
```diff
@@ -0,0 +1,105 @@
+        if (errors && !transport) {
+            console.error('Failed to initialize ledger transport', { errors });
+            throw errors[errors.length - 1];
+        }
```
what if there are no errors and no transport?

MaximusHaximus replied:
In that case, `getSupportedTransports()` will return `null`, which is I think OK for the caller.
Now that you mention it, though, the log in `getSupportedTransport()` for 'transport created!' shouldn't happen if `transport` is null.  Fixed! :)

volovyk-s commented on supportedTransports.js:
```diff
@@ -0,0 +1,105 @@
+        debugLog('Ledger transport created!', transport);
+        return transport;
+    } catch (hwTransportError) {
+        if (hwTransportError.name === 'TransportOpenUserCancelled') {
```
looks like this check does not matter. We will throw `hwTransportError` anyway.

MaximusHaximus replied:
:+1: True indeed; a remnant from the `u2f` fallback logic that I refactored, and already being handled inside `createSupportedTransport()`  now.  Thanks!  Removed this entire catch :)

volovyk-s commented on supportedTransports.js:
```diff
@@ -0,0 +1,105 @@
+        ...(supportWebHid && [{ name: 'WebUSB', createTransport: () => LedgerTransportWebUsb.create() }]),
+
+        // Firefox/Mozilla intend to not support WebHID or WebUSB
+        ...(supportWebHid && [{ name: 'U2F', createTransport: () => LedgerTransportU2F.create() }]),
```
I did not get why do we have `supportWebHid` in all of them. Should we use `supportWebUsb` and `supportHid` instead?

MaximusHaximus replied:
copyPasta fail -- fixed!

volovyk-s commented on supportedTransports.js:
```diff
@@ -0,0 +1,105 @@
+        const { name, createTransport } = supportedTransports[i];
+        debugLog(`Creating ${name} transport`)
+        try {
+            transport = await createTransport();
```
Should we exit this loop if any transport is created successfully?

MaximusHaximus replied:
Already handled -- the for() loop will exit as soon as `transport` is no longer falsy (compound check based on supportedTransports.length && !transport) :+1:

volovyk-s reviewed (approved):
LGTM, I like that we are actually trying to create the `transport` before saying that it will work.

MaximusHaximus commented:
Thanks @volovyk-s <3 Comments resolved!

miraclx commented on supportedTransports.js:
```diff
@@ -0,0 +1,96 @@
+        .catch(() => false);
+}
+
+async function isU2fSupported() {
+    return LedgerTransportWebHid.isSupported()
+        .catch(() => false);
+}
```
Should be:

```suggestion
async function isU2fSupported() {
    return LedgerTransportU2F.isSupported()
        .catch(() => false);
}
```

miraclx commented on supportedTransports.js:
```diff
@@ -0,0 +1,96 @@
+    // We will try each transport we think is supported in the current environment, in order of this array
+    const supportedTransports = [
+        // WebHID is still supported by latest Chrome, so try that first
+        ...(supportWebHid && [{ name: 'WebHID', createTransport: () => LedgerTransportWebHid.create() }]),
+
+        ...(supportWebUsb && [{ name: 'WebUSB', createTransport: () => LedgerTransportWebUsb.create() }]),
+
+        // Firefox/Mozilla intend to not support WebHID or WebUSB
+        ...(supportU2f && [{ name: 'U2F', createTransport: () => LedgerTransportU2F.create() }]),
```
For any non-supported transport, this resolves to `...false` which is invalid logic.

Should be:

```suggestion
        ...(supportWebHid ? [{ name: 'WebHID', createTransport: () => LedgerTransportWebHid.create() }] : []),

        ...(supportWebUsb ? [{ name: 'WebUSB', createTransport: () => LedgerTransportWebUsb.create() }] : []),

        // Firefox/Mozilla intend to not support WebHID or WebUSB
        ...(supportU2f ? [{ name: 'U2F', createTransport: () => LedgerTransportU2F.create() }] : []),
```

miraclx reviewed (commented):
Fixed both issues in https://github.com/near/near-ledger-js/pull/50
